### ISSUE TITLE
Incorrect bundle identifier being used when launching on iOS

### ISSUE THREAD

**Issue by vmaraccini** [flutter/flutter#31037]
Using `flutter run` on an iOS app with custom bundle identifiers (specifically, different bundle identifiers depending on the build configuration) will sometimes fail to launch. The flutter tool builds and installs the app correctly, but then fails to launch it using `xcrun launch ...` because it is supplying the wrong bundle id.

## Steps to Reproduce

The issue manifests itself in two ways:

### 1. Project with custom bundle identifiers on the `pbxproj` definition

1. Create a new project (using `flutter create`)
2. Change the Bundle Identifier of the Debug release. Depending on the order in which the bundle IDs are listed in the `pbxproj`, the flutter tool may fetch the incorrect bundle id (because it always takes the first match):
```
$ cat Runner.xcodeproj/project.pbxproj | grep PRODUCT_BUNDLE_IDENTIFIER

				PRODUCT_BUNDLE_IDENTIFIER = some.bundle.id;
				PRODUCT_BUNDLE_IDENTIFIER = debug.bundle.id;
				PRODUCT_BUNDLE_IDENTIFIER = some.bundle.id;
```

When running `grep -B 30`, we can see the context in which these different `PRODUCT_BUNDLE_IDENTIFIER ` appear:

```
		97C147071CF9000F007C117D /* Release */ = {
			isa = XCBuildConfiguration;
			baseConfigurationReference = 7AFA3C8E1D35360C0083082E /* Release.xcconfig */;
			buildSettings = {
				...
				PRODUCT_BUNDLE_IDENTIFIER = some.bundle.id;
```

In this particular project, Xcode decided to place the `Release` configuration first, hence it showing up first in `grep` and resulting in the flutter tool using `some.bundle.id` to launch the app. 
3.  Use `flutter run`. The app fails to launch in the debug configuration with the following error message: 
```
An error was encountered processing the command (domain=FBSOpenApplicationServiceErrorDomain, code=1):
The request to open "some.bundle.id" failed.
The request was denied by service delegate (SBMainWorkspace) for reason: NotFound ("Application "some.bundle.id" is unknown to FrontBoard").
```
It attempted to launch an app with bundle id `some.bundle.id`,  but it should've been `debug.bundle.id`.

This can be particularly bad if there is another app with the other bundle identifier, in which case an app will launch, but won't be the desired app.

On top of that, the issue is somewhat flaky because Xcode may decide to arrange the `Debug` configuration first in the `pbxproj` file, in which case the app would launch just fine in `Debug` mode, but fail in other modes.

### 2. Project with custom bundle identifiers defined on the `xcconfig` file

1. Create a new project (using `flutter create`)
2. Change the Bundle Identifier to use a value defined in the `xcconfig` file, for example: `$(PRODUCT_BUNDLE_IDENTIFIER)`. 

For example, in the `Debug.xcconfig`, place: `PRODUCT_BUNDLE_IDENTIFIER=debug.bundle.id`
And in the `Release.xcconfig`, add: `PRODUCT_BUNDLE_IDENTIFIER=some.bundle.id`

3. Use `flutter run`. The app fails to launch with the following error:
```
An error was encountered processing the command (domain=FBSOpenApplicationServiceErrorDomain, code=1):
The request to open "${PRODUCT_BUNDLE_IDENTIFIER}" failed.
The request was denied by service delegate (SBMainWorkspace) for reason: NotFound ("Application "${PRODUCT_BUNDLE_IDENTIFIER}" is unknown to FrontBoard").
```

In this case, the flutter tool failed to expand the variable `PRODUCT_BUNDLE_IDENTIFIER` from the `xcconfig` file.
This would be particularly tricky to do as it would require parsing the correct `xcconfig` file and loading the variables from it.

Issue seems to be the same as: https://github.com/flutter/flutter/issues/22587 and https://github.com/flutter/flutter/issues/15567

## Proposal

I propose a solution to this problem where the flutter tool fetches the bundle identifier from the compiled application's `Info.plist`. This is somewhat of a trick, but it avoids the issue of having to parse the `pbxproj` file searching for the correct `PRODUCT_BUNDLE_IDENTIFIER` entry, and also of having to find associated `.xcconfig` files and load their variables. 

See https://github.com/flutter/flutter/pull/31039 for the proposal's implementation.

> **Note**: I have not added the output of `flutter doctor` as it does not seem relevant for this issue, since the tool itself is working fine but is not properly handling the project values.

**Comment by stefandevo** [comment]
Related https://github.com/flutter/flutter/issues/31891

**Comment by jmagman** [comment]
@vmaraccini Can this be closed as fixed by https://github.com/flutter/flutter/pull/31039?

**Comment by stefandevo** [comment]
> @vmaraccini Can this be closed as fixed by #31039?

Yes this is working with the latest bits!

**Comment by jmagman** [comment]
👍
